fix(otlp-transformer): avoid precision loss when converting HrTime to unix nanoseconds

seemk opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

Fixes https://github.com/open-telemetry/opentelemetry-js/issues/2643

## Short description of the changes

Fixes the precision loss when converting `HrTime` to a singular number. The precision loss happened due to `hrTimeToNanoseconds` first multiplying seconds by nanosecond digits, causing the initial number to overflow 53 bits, after which it adds nanoseconds to the number, causing additional errors.

Instead of using the [long.js](https://github.com/dcodeIO/long.js) package, I copied over the relevant parts from this library. long.js does not support CommonJS, only ESM and [it seems enabling `esModuleInterop` is not recommended / accepted](https://github.com/open-telemetry/opentelemetry-js/pull/2467#issuecomment-920314617).

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added
- [x] Documentation has been updated

dyladan commented:
I think this may be a breaking change for the exporter interface? Maybe it's time to start seriously considering a 2.0 release of the SDK since we have quite a few improvements that would be good to make.

seemk commented:
> I think this may be a breaking change for the exporter interface? Maybe it's time to start seriously considering a 2.0 release of the SDK since we have quite a few improvements that would be good to make.

Since the breaking change is in `otlp-transformer` which is both experimental and used internally, I'd keep things as is with regards to this change.

Turns out the amount of tests that needed to be changed was greater than expected, still going over some of them and will try to cut some duplicate code.

seemk commented:
Additionally:
* Removed a bit of dead code in the tests.
* Fixed [opentelemetry-exporter-metrics-otlp-proto/test/OTLPMetricExporter.test.ts](https://github.com/open-telemetry/opentelemetry-js/pull/4062/files#diff-712a67b7aa2e82726dce05de9454b18060ed032c83e82a6e9febaa44cba7b1a4) and [opentelemetry-exporter-metrics-otlp-grpc/test/OTLPMetricExporter.test.ts](https://github.com/open-telemetry/opentelemetry-js/pull/4062/files#diff-027af27e5e7182406fd3a331263372057498022868949b788f2349fd77329ef5) to actually test against the ground truth. It was comparing the parsed timestamps against the parsed timestamps. The data was decoded and then cast to a mismatching type, in the case of timestamps the actual parsed data had string timestamps, but this was never reflected in the OTLP types.

pichlermarc reviewed (commented):
> > I think this may be a breaking change for the exporter interface? Maybe it's time to start seriously considering a 2.0 release of the SDK since we have quite a few improvements that would be good to make.
> 
> Since the breaking change is in `otlp-transformer` which is both experimental and used internally, I'd keep things as is with regards to this change.

I agree, this change should only be a problem to direct users of `@opentelemetry/otlp-transformer`, and it's documented that this package is only intended for internal use. The exporters are pinned to the exact version, so it's fair to assume that this won't break actual end-users. :slightly_smiling_face: 

> Turns out the amount of tests that needed to be changed was greater than expected, still going over some of them and will try to cut some duplicate code.

Yes, this has been a pain point for a while. They have grown historically and really need some refactoring. :disappointed: 

> Additionally:
> 
>     * Removed a bit of dead code in the tests.
> 
>     * Fixed [opentelemetry-exporter-metrics-otlp-proto/test/OTLPMetricExporter.test.ts](https://github.com/open-telemetry/opentelemetry-js/pull/4062/files#diff-712a67b7aa2e82726dce05de9454b18060ed032c83e82a6e9febaa44cba7b1a4) and [opentelemetry-exporter-metrics-otlp-grpc/test/OTLPMetricExporter.test.ts](https://github.com/open-telemetry/opentelemetry-js/pull/4062/files#diff-027af27e5e7182406fd3a331263372057498022868949b788f2349fd77329ef5) to actually test against the ground truth. It was comparing the parsed timestamps against the parsed timestamps. The data was decoded and then cast to a mismatching type, in the case of timestamps the actual parsed data had string timestamps, but this was never reflected in the OTLP types.

Thanks! :rocket: I've added one blocking comment about the code adapted from long.js - then I think this PR is good to go. 
Waiting for approval by @dyladan before merging though to ensure all concerns are addressed :slightly_smiling_face:

seemk commented on experimental/packages/otlp-transformer/src/common/unsigned_long.ts:
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright The OpenTelemetry Authors
```
Thanks!

Ankcorn commented:
Hey @seemk , just a heads up this was a breaking change for us and ingesting open telemetry spans. It caused a little bit of confusion but we did figure out what was going on eventually :sweat_smile: 

We have our own custom receptor written in typescript

Are we doing something wrong/unsafe and is there an official package for deserializing spans?

seemk commented:
> Hey @seemk , just a heads up this was a breaking change for us and ingesting open telemetry spans. It caused a little bit of confusion but we did figure out what was going on eventually :sweat_smile: 
> 
> We have our own custom receptor written in typescript
> 
> Are we doing something wrong/unsafe and is there an official package for deserializing spans? 

What does the custom receiver do? Do you have your own protobuf decoder?

karanssj4 commented:
@seemk we faced a similar issue, timestamps started coming up as an object 
```
hypertest-logger-consumer-1  |       startTimeUnixNano: {
hypertest-logger-consumer-1  |         low: 2022226944,
hypertest-logger-consumer-1  |         high: 395220561
hypertest-logger-consumer-1  |       },
hypertest-logger-consumer-1  |       endTimeUnixNano: {
hypertest-logger-consumer-1  |         low: 2023255286,
hypertest-logger-consumer-1  |         high: 395220561
hypertest-logger-consumer-1  |       }

hypertest-logger-consumer-1  | Argument `startTimeUnixNano`: Invalid value provided. Expected BigInt, provided Object.

```

karanssj4 commented:
@Ankcorn can confirm protobuf wasn't affected

lingyan commented:
BTW, there is a detailed report of the breakage in this issue: https://github.com/open-telemetry/opentelemetry-js/issues/4202

cristianmadularu commented:
Hi guys, I'm not sure if this change was meant to have external visibility, but the timestamps being sent by a regular web application are also being sent at JSON objects now breaking certain providers ingressing this data (e.g.. Uptrace).
Please see https://github.com/open-telemetry/opentelemetry-js/issues/4216

polRk commented:
I think, we should revert this commit. I can not send any data 
```js
{"stack":"OTLPExporterError: Bad Request\n    at IncomingMessage.<anonymous> (/home/polyakov/sberinfra-cloud-console/packages/telemetry/node_modules/@opentelemetry/otlp-exporter-base/build/src/platform/node/util.js:103:39)\n    at IncomingMessage.emit (node:events:526:35)\n    at endReadableNT (node:internal/streams/readable:1376:12)\n    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)","message":"Bad Request","name":"OTLPExporterError","data":"{\"code\":3,\"message\":\"ReadUint64: unsupported value type, error found in #10 byte of ...|UnixNano\\\":{\\\"low\\\":194|..., bigger context ...|\\\":\\\"@webui/graphql\\\"},\\\"logRecords\\\":[{\\\"timeUnixNano\\\":{\\\"low\\\":1948385984,\\\"high\\\":395403571},\\\"observedTimeU|...\"}","code":"400"}
```

sergeynechaev commented:
Hi guys! I can confirm that this change also broke sending browser events to Honeycomb.io. Their API `https://api.honeycomb.io/v1/traces` expects string or number in the `startTimeUnixNano` and `endTimeUnixNano`, but get the IFixed64 object instead and fails with the `"failed to parse OTLP request body"` error.

```
export interface IFixed64 {
    low: number;
    high: number;
}
```
